packages/ckeditor5-engine/src/model/operation/rootoperation.ts

```diff
@@ -0,0 +1,154 @@
+/**
+ * @license Copyright (c) 2003-2023, CKSource Holding sp. z o.o. All rights reserved.
+ * For licensing, see LICENSE.md or https://ckeditor.com/legal/ckeditor-oss-license
+ */
+
+/**
+ * @module engine/model/operation/rootoperation
+ */
+
+import Operation from './operation';
+
+import type Document from '../document';
+import { CKEditorError } from '@ckeditor/ckeditor5-utils';
+
+/**
+ * Operation that creates or removes a root element.
+ */
+export default class RootOperation extends Operation {
+	/**
+	 * Root name to create or remove.
+	 */
+	public readonly rootName: string;
+
+	/**
+	 * Root element to create or remove.
+	 */
+	public readonly elementName: string;
+
+	/**
+	 * Specifies whether the operation adds (`true`) or removes the root (`false`).
+	 */
+	public readonly isAdd: boolean;
+
+	/**
+	 * Document for which the root should be created.
+	 */
+	private readonly _document: Document;
+
+	/**
+	 * Creates an operation that creates or removes a root element.
+	 *
+	 * @param rootName Root name.
+	 * @param elementName Root element name.
+	 * @param isAdd Specifies whether the operation adds (`true`) or removes the root (`false`).
+	 * @param document Document for which the root should be created.
+	 * @param baseVersion Document {@link module:engine/model/document~Document#version} on which operation
+	 * can be applied or `null` if the operation operates on detached (non-document) tree.
+	 */
+	constructor(
+		rootName: string,
+		elementName: string,
+		isAdd: boolean,
+		document: Document,
+		baseVersion: number | null
+	) {
+		super( baseVersion );
+
+		this.rootName = rootName;
+		this.elementName = elementName;
+		this.isAdd = isAdd;
+		this._document = document;
+
+		// Make sure that the root exists ASAP, this is important for RTC.
+		// If the root was dynamically added, there will be more operations that operate on/in this root.
+		// These operations will require root element instance (in operation property or in position instance).
+		// If the root is not created ahead of time, instantiating such operations may fail.
+		if ( !this._document.getRoot( this.rootName ) ) {
+			const root = this._document.createRoot( this.elementName, this.rootName );
+
+			root._isAttached = false;
+		}
+	}
+
+	public override get type(): 'addRoot' | 'removeRoot' {
```

Comment:
Missing doclets for methods.

Reply:
I wasn't sure if we actually need to add these if they are just

```
/**
 * @inheritDoc
 */
```